compiler/src/org.graalvm.compiler.truffle.runtime/src/org/graalvm/compiler/truffle/runtime/debug/disassembler/MachineCode.java

```diff
@@ -0,0 +1,148 @@
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+/**
+ * Represents machine code for disassembly purposes.
+ */
+public class MachineCode {
+
+    private final long address;
+    private final long headerLength;
+    private final byte[] code;
+    private final Infopoint[] infopoints;
+
+    public static MachineCode read(long address, long headerLength, int size, Infopoint[] infopoints) {
+        return new MachineCode(address, headerLength, readCode(address, size), infopoints);
+    }
+
+    private MachineCode(long address, long headerLength, byte[] code, Infopoint[] infopoints) {
+        this.address = address;
+        this.headerLength = headerLength;
+        this.code = code;
+        this.infopoints = infopoints;
+    }
+
+    public int getLength() {
+        return code.length;
+    }
+
+    public String disassemble(DisassemblyFormatType disassemblyFormat) throws IOException {
+        return createDisassembler(disassemblyFormat).disassemble(new MachineCodeAccessor() {
+
+            @Override
+            public long getAddress() {
+                return address;
+            }
+
+            @Override
+            public long getHeaderLength() {
+                return headerLength;
+            }
+
+            @Override
+            public int getLength() {
+                return code.length;
+            }
+
+            @Override
+            public byte getByte(int n) {
+                return code[n];
+            }
+
+            @Override
+            public byte[] getBytes() {
+                return Arrays.copyOf(code, code.length);
+            }
+
+            @Override
+            public Stream<Infopoint> getInfopoints() {
+                return Arrays.stream(infopoints);
+            }
+
+            @Override
+            public String fileName(String extension) {
+                return String.format("truffle_compiled_code_%d_0x%x_%d%s", getPid(), address, System.nanoTime(), extension);
+            }
+
+        });
+    }
+
+    private Disassembler createDisassembler(DisassemblyFormatType disassemblyFormat) {
+        switch (disassemblyFormat) {
+            case HEX:
+                return new HexDisassembler();
+            case RAW:
+                return new RawDisassembler();
+            case ELF:
+                return new ElfDisassembler();
+            case OBJDUMP:
+                return new ObjdumpDisassembler();
+            default:
+                throw new UnsupportedOperationException();
+        }
+    }
+
+    private static long getPid() {
+        try {
+            final String info = java.lang.management.ManagementFactory.getRuntimeMXBean().getName();
+            return Long.parseLong(info.split("@")[0]);
+        } catch (Exception e) {
+            return 0;
+        }
+    }
+
+    private static byte[] readCode(long address, int size) {
+        final byte[] bytes = new byte[size];
```

Comment:
On HotSpot it may be better to use `InstalledCode.getCode`. You will need to add the `getCode` method to `OptimizedCallTarget`. In `HotSpotOptimizedCallTarget` you can delegate to `InstalledCode.getCode`. In `SubstrateOptimizedCallTarget` you will need the unsafe to read the code as you do.